imap-proto/src/parser/rfc2087.rs

```diff
@@ -0,0 +1,273 @@
+//!
+//! https://tools.ietf.org/html/rfc2087
+//!
+//! IMAP4 QUOTA extension
+//!
+
+use std::borrow::Cow;
+
+use nom::{
+    branch::alt,
+    bytes::streaming::{tag, tag_no_case},
+    character::streaming::space1,
+    combinator::map,
+    multi::many0,
+    multi::separated_list0,
+    sequence::{delimited, preceded, tuple},
+    IResult,
+};
+
+use crate::parser::core::astring_utf8;
+use crate::types::*;
+
+use super::core::number_64;
+
+/// 5.1. QUOTA Response
+/// ```ignore
+/// quota_response  ::= "QUOTA" SP astring SP quota_list
+/// ```
+pub(crate) fn quota(i: &[u8]) -> IResult<&[u8], Response> {
+    let (rest, (_, _, root_name, _, resources)) = tuple((
+        tag_no_case("QUOTA"),
+        space1,
+        map(astring_utf8, Cow::Borrowed),
+        space1,
+        quota_list,
+    ))(i)?;
+
+    Ok((
+        rest,
+        Response::Quota(Quota {
+            root_name,
+            resources,
+        }),
+    ))
+}
+
+/// ```ignore
+/// quota_list  ::= "(" #quota_resource ")"
+/// ```
+pub(crate) fn quota_list(i: &[u8]) -> IResult<&[u8], Vec<QuotaResource>> {
+    delimited(tag("("), separated_list0(space1, quota_resource), tag(")"))(i)
+}
+
+/// ```ignore
+/// quota_resource  ::= atom SP number SP number
+/// ```
+pub(crate) fn quota_resource(i: &[u8]) -> IResult<&[u8], QuotaResource> {
+    let (rest, (name, _, usage, _, limit)) = tuple((
+        quota_resource_name,
+        tag(" "),
+        number_64,
+        tag(" "),
+        number_64,
+    ))(i)?;
+
+    Ok((rest, QuotaResource { name, usage, limit }))
+}
+
+pub(crate) fn quota_resource_name(i: &[u8]) -> IResult<&[u8], QuotaResourceName> {
+    alt((
+        map(tag_no_case("STORAGE"), |_| QuotaResourceName::Storage),
+        map(tag_no_case("MESSAGE"), |_| QuotaResourceName::Message),
+        map(map(astring_utf8, Cow::Borrowed), QuotaResourceName::Atom),
+    ))(i)
+}
+
+/// 5.2. QUOTAROOT Response
+/// ```ignore
+/// quotaroot_response ::= "QUOTAROOT" SP astring *(SP astring)
+/// ```
+pub(crate) fn quota_root(i: &[u8]) -> IResult<&[u8], Response> {
+    let (rest, (_, _, mailbox_name, quota_root_names)) = tuple((
+        tag_no_case("QUOTAROOT"),
+        space1,
+        map(astring_utf8, Cow::Borrowed),
+        many0(preceded(space1, map(astring_utf8, Cow::Borrowed))),
+    ))(i)?;
+
+    Ok((
+        rest,
+        Response::QuotaRoot(QuotaRoot {
+            mailbox_name,
+            quota_root_names,
+        }),
+    ))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use assert_matches::assert_matches;
+    use std::borrow::Cow;
+
+    #[test]
+    fn test_quota() {
+        assert_matches! (quota(b"QUOTA \"\" (STORAGE 10 512)") ,
```

Comment:
Nit: here and elsewhere, please eliminate the space after `assert_matches!`, align the first argument on the next line and eliminate the space before the comma.